### PR TITLE
Expose `delta.default-reader-version` and `delta.default-writer-version` delta lake configuration properties

### PR DESCRIPTION
**Pull request by krvikash**
## Description
Follow up of https://github.com/trinodb/trino/pull/16165#discussion_r1113247631

## Release notes

(X) Release notes are required, with the following suggested text:

```markdown
# Delta Lake
* Support setting default reader and writer versions using the `delta.default-reader-version` and 
  `delta.default-writer-version` config properties. ({issue}`16208`)
```

### CODE REVIEW

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeConfig.java [hunk 1]
```diff
@@ -77,6 +77,8 @@
     private boolean uniqueTableLocation = true;
     private boolean legacyCreateTableWithExistingLocationEnabled;
     private boolean registerTableProcedureEnabled;
+    private int readerVersion = 1;
+    private int writerVersion = 2;
```
Comment (ebyhr) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeConfig.java:81]: It would be nice to deny setting unsupported versions. You can refer to `MySqlJdbcConfig`.

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeConfig.java [hunk 2]
```diff
@@ -475,4 +477,30 @@ public DeltaLakeConfig setRegisterTableProcedureEnabled(boolean registerTablePro
...
+    public DeltaLakeConfig setWriterVersion(int writerVersion)
+    {
+        this.writerVersion = writerVersion;
+        return this;
+    }
+
+    public int getWriterVersion()
+    {
+        return writerVersion;
+    }
```
Comment (ebyhr) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeConfig.java:505]: Replace getter and setter order for consistency with other properties.

Comment (krvikash) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeConfig.java:505]: setter-getter is inconsistent in this file (setter-getter then getter-setter). What is the ideal way?

Comment (ebyhr) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeConfig.java:505]: At least, my IntelliJ generates getter-setter.

**Comment by krvikash** [comment]
Thank @ebyhr for the review. Addressed comments.

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeConfig.java [hunk 3]
```diff
@@ -134,8 +140,35 @@ public void testExplicitPropertyMappings()
...
+
+    @Test
+    public void testReaderWriterVersion()
+    {
+        assertFalse(isValidReaderVersion(0));
+        assertTrue(isValidReaderVersion(2));
+        assertFalse(isValidReaderVersion(3));
+        assertFalse(isValidWriterVersion(1));
+        assertTrue(isValidWriterVersion(2));
+        assertFalse(isValidWriterVersion(5));
```
Comment (ebyhr) [inline plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeConfig.java:158]: I would recommend using `io.airlift.testing.ValidationAssertions#assertFailsValidation`.

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java [hunk 4]
```diff
@@ -1782,8 +1783,8 @@ private ProtocolEntry getProtocolEntry(Map<String, Object> properties)
             }
         }
         return new ProtocolEntry(
-                readerVersion.orElse(MIN_READER_VERSION),
-                writerVersion.orElse(MIN_WRITER_VERSION));
+                readerVersion.orElse(deltaLakeConfig.getReaderVersion()),
+                writerVersion.orElse(deltaLakeConfig.getWriterVersion()));
```
Comment (ebyhr) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java:1787]: I couldn't understand the intention of this change. Did you want to set default reader and writer versions of `DeltaLakeConfig`?

Comment (krvikash) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java:1787]: Yes I want to set default reader and writer version.

Comment (ebyhr) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java:1787]: Why do you initialize `DeltaLakeConfig` at L1767? It ignores configured values by users.

Comment (krvikash) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java:1787]: My intension was, If `reader_version` and `writer_version` are overridden from `CREATE TABLE ...` or `ALTER TABLE ...` statement then use it, otherwise use it from the delta config properties.

I got the problem over here. I have to use configured DeltaLakeConfig and not with default values only.

Comment (krvikash) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java:1787]: I see that we get `DeltaLakeConfig` properties in `DeltaLakeMetadataFactory` and pass them to `DeltaLakeMetadata` as arguments. This way it will keep on increasing the number of arguments if we want to use more and more `DeltaLakeConfig` properties in `DeltaLakeMetadata`.

https://github.com/trinodb/trino/blob/c12f3a69e3eed14ccff715fbaa6d9fcf6b471ef3/plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadataFactory.java#L97-L102

WDYT, If we refactor (in a separate commit) DeltaLakeMetadataFactory to pass `DeltaLakeConfig` as an argument to  `DeltaLakeMetadata`? In growing use of `DeltaLakeConfig` properties in `DeltaLakeMetadata` will ensure no change in `DeltaLakeMetadataFactory`.

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeConfig.java [hunk 5]
```diff
@@ -475,4 +487,53 @@ public DeltaLakeConfig setRegisterTableProcedureEnabled(boolean registerTablePro
...
+    public boolean isValidReaderVersion()
+    {
+        return isValidVersion("delta.reader-version", readerVersion, MIN_READER_VERSION, MAX_READER_VERSION);
+    }
+
+    @AssertTrue(message = "Unsupported writer version for Delta Lake connector")
+    public boolean isValidWriterVersion()
+    {
+        return isValidVersion("delta.writer-version", writerVersion, MIN_WRITER_VERSION, MAX_WRITER_VERSION);
+    }
```
Comment (pajaks) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeConfig.java:527]: Are those functions used only in tests? If yes then they should be probably moved there, or maybe they should be used in production code?

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/BaseDeltaLakeMinioConnectorTest.java [hunk 6]
```diff
@@ -862,6 +865,197 @@ public void testThatEnableCdfTablePropertyIsShownForCtasTables()
...
+    {
+        String tableName = "test_alter_table_upgrade_reader_writer_version_" + randomNameSuffix();
+
+        assertUpdate(format("CREATE TABLE %s.%s (a_number INT)", SCHEMA, tableName));
+
+        assertUpdate("ALTER TABLE " + tableName + " SET PROPERTIES reader_version = 2");
+        assertThatShowCreateTable(tableName, ".*reader_version = 2.*");
+
+        assertUpdate("ALTER TABLE " + tableName + " SET PROPERTIES writer_version = 3");
+        assertThatShowCreateTable(tableName, ".*writer_version = 3.*");
```
Comment (pajaks) [inline plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/BaseDeltaLakeMinioConnectorTest.java:1035]: Can we also somehow check if transaction log is updated with new entry?

**Review by ebyhr — CHANGES_REQUESTED** [review]

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeConfig.java (hunk 5 above)
Comment (pajaks) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeConfig.java:527]: Ok. This annotation is used during object validation. It is clear after test changes.

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/BaseDeltaLakeMinioConnectorTest.java (hunk 6 above)
Comment (krvikash) [reply plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/BaseDeltaLakeMinioConnectorTest.java:1035]: `SHOW CREATE TABLE ...` actually does the same thing. It reads the transaction log file, parses and sets the `protocolEntry` and other objects. So I think it is safe to assume that the version returned from `SHOW CREATE TABLE` is the same as of transaction log.

https://github.com/trinodb/trino/blob/master/plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/checkpoint/TransactionLogTail.java#L138

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java [hunk 7]
```diff
@@ -321,8 +318,7 @@
     private final AtomicReference<Runnable> rollbackAction = new AtomicReference<>();
     private final DeltaLakeRedirectionsProvider deltaLakeRedirectionsProvider;
     private final ExtendedStatisticsAccess statisticsAccess;
-    private final boolean deleteSchemaLocationsFallback;
-    private final boolean useUniqueTableLocation;
+    private final DeltaLakeConfig deltaLakeConfig;
```
[the remark this answers is not in the document]
Comment (krvikash) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java]: Thanks for the reference. Reverted.

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeConfig.java [hunk 8]
```diff
@@ -475,4 +487,53 @@ public DeltaLakeConfig setRegisterTableProcedureEnabled(boolean registerTablePro
...
+
+    @Config("delta.writer-version")
+    @ConfigDescription("The writer version used by tables created by Trino")
+    public DeltaLakeConfig setWriterVersion(int writerVersion)
+    {
+        this.writerVersion = writerVersion;
+        return this;
+    }
+
+    @AssertTrue(message = "Unsupported reader version for Delta Lake connector")
```
[the remark this answers is not in the document]
Comment (krvikash) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeConfig.java]: Used `@Min` and `@Max` annotations. This will not require checking version explicitly.

**Review by ebyhr — COMMENTED** [review]
Commit granularity seems wrong. Adding config properties without actual usage doesn't make sense. I would recommend squashing three commits into one.

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeConfig.java [hunk 9]
```diff
@@ -475,4 +485,34 @@ public DeltaLakeConfig setRegisterTableProcedureEnabled(boolean registerTablePro
...
     }
+
+    @Min(value = MIN_READER_VERSION, message = "Must be in between " + MIN_READER_VERSION + " and " + MAX_READER_VERSION)
+    @Max(value = MAX_READER_VERSION, message = "Must be in between " + MIN_READER_VERSION + " and " + MAX_READER_VERSION)
+    public int getReaderVersion()
+    {
+        return readerVersion;
+    }
+
+    @Config("delta.reader-version")
```
Comment (ebyhr) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeConfig.java:496]: `delta.default-reader-version` might be better as `delta.default-checkpoint-writing-interval`. What do you think about it?

**Review by ebyhr — APPROVED** [review]

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeConfig.java (hunk 9 above)
Comment (krvikash) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeConfig.java]: Make sense. I have renamed the property.

**Comment by ebyhr** [comment]
/test-with-secrets sha=0326b9e5680df72d04db320e6c223590d609c3cf

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeConfig.java [hunk 10]
```diff
@@ -48,6 +49,13 @@
     @VisibleForTesting
     static final DataSize DEFAULT_DATA_FILE_CACHE_SIZE = DataSize.succinctBytes(Math.floorDiv(Runtime.getRuntime().maxMemory(), 10L));
 
+    // The required reader and writer versions used by tables created by Trino
```
Comment (alexjo2144) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeConfig.java:52]: Now that we have CDC this comment isn't really right, let's remove it. You could rephrase it to say that tables cannot be created with versions lower than them.

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeConfig.java [hunk 11]
```diff
@@ -475,4 +485,34 @@ public DeltaLakeConfig setRegisterTableProcedureEnabled(boolean registerTablePro
...
+
+    @Min(value = MIN_READER_VERSION, message = "Must be in between " + MIN_READER_VERSION + " and " + MAX_READER_VERSION)
+    @Max(value = MAX_READER_VERSION, message = "Must be in between " + MIN_READER_VERSION + " and " + MAX_READER_VERSION)
+    public int getDefaultReaderVersion()
+    {
+        return defaultReaderVersion;
+    }
+
+    @Config("delta.default-reader-version")
+    @ConfigDescription("The reader version used by tables created by Trino")
```
Comment (alexjo2144) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeConfig.java:497]: "by Trino" is a little redundant I think :)
```suggestion
    @ConfigDescription("The default reader version used by new tables")
```

**Review by alexjo2144 — APPROVED** [review]
Nitpicks
